### PR TITLE
Docs infrastructure: swap out the `nb2plot` sphinx extension with `myst-nb`

### PR DESCRIPTION
**Pull request by rossbar**
Both the tutorial and the introduction rely on executable code embedded in the corresponding document being run when the docs are being built. This is currently done with the `.. nbplot` directive, provided by the nb2plots sphinx extension. There is some overlap of functionality with this extension and the newer `myst-nb` sphinx extension from the executable books project, which also supports markup documents with executable cells.

I thought I'd give `myst-nb` a try as an alternative to `.. nbplot` to get a sense of whether this is a viable alternative for the documents. After a bit of configuring I think it's quite possible and actually has a really nice workflow with sphinx. The one major change is that using `myst-nb` would *require* the tutorial and introduction to be written in myst-markdown instead of rst. This is because, AFAICT, the `{code-cell}` used by `myst-nb` to denote executable cells is not a generic sphinx directive, and thus is not supported in the rst files.

IMO it's not pressing and there aren't any *major* improvements that motivate the change: it gets rid of a few baked-in sphinx warnings from `nb2plots` and also provides execution caching, but that doesn't have a huge effect in this case since it's only the two files that are executed and neither of them has a very long run time. I do think it's a viable alternative however and would be interested in discussing it if anyone's interested!

### CODE REVIEW

**Comment by rossbar** [comment]
Marking as draft for now since I'm mostly interested in starting a discussion!

**Comment by rossbar** [comment]
FWIW the failures are related to building the pdf. The doc preview via the build artifact still works, despite the red x. PDF build problems have been reported upstream, see executablebooks/MyST-NB#375

**Comment by rossbar** [comment]
> I also wanted to see if it's possible to start using jupyterlite to setup our tutorial with in browser interactivity. It still requires a .ipynb notebook for now but myst-nb is closer to a notebook (in my head) than nb2plots.

Yeah this would be really cool :+1: Does it work with .md-formatted notebooks if jupytext is installed in the environment (assuming it's possible to do so - I'm not sure "environment" makes sense in this context)?

> I would love to see this move forward, fixed the conflicts to get the CI working again.

Me too - the main blocker is that myst-nb doesn't work 100% with the sphinx latex builder, but if we remove the pdf generation from our CI runs (which has been discussed in the past) then I think that blocker is removed. FWIW the released version of nb2plots no longer works with Python 3.10 (there's a patch on `master` but it hasn't been released yet) which is another motivation to make the switch in my mind!

**Comment by rossbar** [comment]
Okay updates here!

The good news: the latex builder problems seem to be resolved. I no longer see issues locally running `make dist` with myst-nb v0.17.2.

I've also tested the doc build locally with both sphinx 6.2 and sphinx 7.0.1 and everything is working, so this PR resolves all of the issues related to depending on `nb2plot`.

The bad news is that many of the myst tools (primarily `myst-nb` and `myst-parser`, at time of writing) are still pinned to sphinx<6, which would force us to either manually work around this in CI, or temporarily downgrade in order to accommodate this restriction (see 1144f27).

**Comment by rossbar** [comment]
Good news - with the release of `myst-nb` 1.0 there are no longer any myst-specific sphinx upperbounds. That removes the last real blocker from this moving forward. I've rebased and this is now ready for review!

File: doc/conf.py [hunk 1]
```diff
@@ -77,8 +77,9 @@
 # The encoding of source files.
 source_encoding = "utf-8"
 
-# Do not include release announcement template
-exclude_patterns = ["release/release_template.rst"]
+# Items to exclude during source collection, including release announcement template,
+# build outpus and README
```
Comment (dschult) [inline doc/conf.py:81]:
```suggestion
# build output and README
```

File: doc/reference/introduction.md [hunk 2]
```diff
@@ -0,0 +1,351 @@
+---
```
Comment (dschult) [inline doc/reference/introduction.md:1]: Would it be better for blame if this new file was first moved from introduction.md and then changed? Or maybe the blame tracker doesn't go across renamed files....  not a blocker -- just wondered...

**Review by dschult — APPROVED** [review]
This looks good!  
It's hard to tell what changes were made in the shift from rst to md. 
How hard was it to make that switch? What did you have to change?
(and below I talk about whether there is a way to get github/blame to track the changes even though the file name changes.

Anyway, this looks good to me -- I don't feel qualified in terms of notebooks to tell whether it does what we want or not. So I'll let @MridulS do the official approval.
Thanks for keeping track of this!!

**Comment by rossbar** [comment]
>  How hard was it to make that switch? What did you have to change?

I used an automatic converter [`rst2myst`](https://rst-to-myst.readthedocs.io/en/latest/) to do the conversion, then made a few tweaks to clean up a few issues from the autoconverter (see e.g. [0d3d48b](https://github.com/networkx/networkx/pull/5221/commits/0d3d48b19effb1f8e4824c48f088f75a6660a393)). It's a fairly painless process and *in principle* the only thing that changed was the formatting - the content of the files should remain unchanged.

>  It's hard to tell what changes were made in the shift from rst to md.

Yeah unfortunately this is the big downside of converting from one markup text to another. The most important consideration is that the content is the same, followed by ensuring that docs render correctly in html. Unfortunately the diffs aren't particularly helpful and I can't think of a way to improve the situation - but I'm willing to try anything so LMK if you have ideas!

> Would it be better for blame if this new file was first moved from introduction.md and then changed? Or maybe the blame tracker doesn't go across renamed files.... not a blocker -- just wondered...

Good question - I will give this a try on a separate branch and see if it makes a difference for blame!

**Comment by rossbar** [comment]
> Would it be better for blame if this new file was first moved from introduction.md and then changed?

I'm glad you caught this - there is indeed a formula that can preserve the blame in the GH interface! See e.g. [here](https://github.com/rossbar/networkx/blame/mystnb-preserve-blame/doc/tutorial.md).

I've applied the procedure to the tutorial, but I still need to transfer over the introduction as well. Rather than try to munge/force-push over the history with what's already here, I think it'd be much easier to open a new PR. I will close this one now, and open a new PR with a cleaner history + cross-file blame preservation as soon as I finish converting! The content/changes will be the same, just applied in a different fashion to better preserve blame as suggested. Thanks again for the idea @dschult !

**Pull request closed without merging**